### PR TITLE
Implement tagged template literals

### PR DESCRIPTION
**Pull request by rerobika**
Missing features: snapshot support

JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

**Comment by rerobika** [comment]
The patch is ready for the review. Due to the template object caching currently this feature is unsupported by the snapshot tool. I'm planning to add support for it, but due to the already enormous size of the patch, I'm going to do it in a followup PR.

File: tests/unit-core/test-snapshot.c [hunk 1]
```diff
@@ -223,7 +223,7 @@ main (void)
     /* Check the snapshot data. Unused bytes should be filled with zeroes */
     const uint8_t expected_data[] =
     {
-      0x4A, 0x52, 0x52, 0x59, 0x22, 0x00, 0x00, 0x00,
+      0x4A, 0x52, 0x52, 0x59, 0x23, 0x00, 0x00, 0x00,
```
Comment (LaszloLango) [inline tests/unit-core/test-snapshot.c]: What has changed if this feature does not work with snapshots?

Comment (rerobika) [reply tests/unit-core/test-snapshot.c]: New cbc opcodes has been added.

File: jerry-core/parser/js/js-parser-expr.c [hunk 2]
```diff
@@ -1189,30 +1190,114 @@ parser_parse_template_literal (parser_context_t *context_p) /**< context */
...
+    coll_p = ECMA_GET_INTERNAL_VALUE_POINTER (ecma_collection_t, context_p->tagged_template_literal_cp);
+    if (coll_p->item_count > CBC_MAXIMUM_BYTE_VALUE)
+    {
+      parser_raise_error (context_p, PARSER_ERR_ARGUMENT_LIMIT_REACHED);
+    }
+  }
+  uint32_t tagged_id = coll_p->item_count;
+  uint32_t prop_idx = 0;
+  ecma_object_t *template_obj_p = ecma_op_new_array_object (0);
+  ecma_object_t *raw_strings_p = ecma_op_new_array_object (0);
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-expr.c]: Set this as raw property here, not at the end.

File: jerry-core/parser/js/js-parser-internal.h [hunk 3]
```diff
@@ -631,10 +636,13 @@ bool lexer_check_arrow (parser_context_t *context_p);
 bool lexer_check_arrow_param (parser_context_t *context_p);
 bool lexer_check_yield_no_arg (parser_context_t *context_p);
 #endif /* ENABLED (JERRY_ES2015) */
-void lexer_parse_string (parser_context_t *context_p);
+void lexer_parse_string (parser_context_t *context_p, lexer_string_options_t opts);
 void lexer_expect_identifier (parser_context_t *context_p, uint8_t literal_type);
 void lexer_scan_identifier (parser_context_t *context_p, uint32_t ident_opts);
-void lexer_convert_ident_to_cesu8 (uint8_t *destination_p, const uint8_t *source_p, prop_length_t length);
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-internal.h]: Please revert this change.

File: jerry-core/parser/js/js-parser-mem.c [hunk 4]
```diff
@@ -88,6 +88,19 @@ void parser_free_local (void *ptr, /**< pointer to free */
   jmem_heap_free_block (ptr, size);
 } /* parser_free_local */
 
+/**
+ * Free the dynamically allocated buffer stored in the context
+ */
+void parser_free_allocated_buffer (parser_context_t *context_p) /**< context */
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-mem.c]: Newline after void.

File: jerry-core/parser/js/js-parser.c [hunk 5]
```diff
@@ -2482,9 +2516,29 @@ parser_raise_error (parser_context_t *context_p, /**< context */
       parser_free_jumps (saved_context_p->last_statement);
     }
 
+#if ENABLED (JERRY_ES2015)
+    if (saved_context_p->tagged_template_literal_cp != JMEM_CP_NULL)
+    {
+      ecma_collection_t *coll_p = ECMA_GET_INTERNAL_VALUE_POINTER (ecma_collection_t,
+                                                                   saved_context_p->tagged_template_literal_cp);
+
+      ecma_collection_free (coll_p);
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser.c]: This local value does not needed.

File: jerry-core/parser/js/js-parser.c [hunk 6]
```diff
@@ -2482,9 +2516,29 @@ parser_raise_error (parser_context_t *context_p, /**< context */
...
     saved_context_p = saved_context_p->prev_context_p;
   }
 
+#if ENABLED (JERRY_ES2015)
+  if (context_p->tagged_template_literal_cp != JMEM_CP_NULL)
+  {
+    ecma_collection_t *coll_p = ECMA_GET_INTERNAL_VALUE_POINTER (ecma_collection_t,
+                                                                 context_p->tagged_template_literal_cp);
+
+    ecma_collection_free (coll_p);
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser.c]: Ditto.

File: jerry-core/vm/vm.c [hunk 7]
```diff
@@ -2096,7 +2096,40 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
 
           goto error;
         }
-#endif /* ENABLED (JERRY_ES2015) */
+        case VM_OC_TEMPLATE_OP:
+        {
+          ecma_string_t *lhs_p = ecma_op_to_string (left_value);
```
Comment (zherczeg) [inline jerry-core/vm/vm.c]: `lefthandside_p`

File: jerry-core/vm/vm.c [hunk 8]
```diff
@@ -2096,7 +2096,40 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
...
+          ecma_deref_ecma_string (rhs_p);
+
+          *stack_top_p++ = ecma_make_string_value (result_str_p);
+          goto free_both_values;
+        }
+        case VM_OC_GET_TEMPLATE_OBJECT:
+        {
+          uint8_t tagged_idx = *byte_code_p++;
+          ecma_collection_t *coll_p = ecma_compiled_code_get_tagged_template_collection (bytecode_header_p);
+          JERRY_ASSERT (tagged_idx < coll_p->item_count);
```
Comment (zherczeg) [inline jerry-core/vm/vm.c]: coll_p

File: jerry-core/parser/js/js-scanner.c [hunk 9]
```diff
@@ -2695,6 +2710,12 @@ scanner_scan_all (parser_context_t *context_p, /**< context */
         {
           if (scanner_scan_post_primary_expression (context_p, &scanner_context, type, stack_top))
           {
+#if ENABLED (JERRY_ES2015)
+            if (JERRY_UNLIKELY (context_p->source_p[-1] == LIT_CHAR_GRAVE_ACCENT))
+            {
+              continue;
+            }
+#endif /* ENABLED (JERRY_ES2015) */
```
Comment (zherczeg) [inline jerry-core/parser/js/js-scanner.c]: This is not needed.

File: tests/jerry/es2015/template_string.js [hunk 10]
```diff
@@ -90,3 +90,15 @@ must_throw ("`${}`");
...
+var cr = eval("`a" + String.fromCharCode(13) + "b`");
+var lf = eval("`a" + String.fromCharCode(10) + "b`");
+var crlf = eval("`a" + String.fromCharCode(13,10) + "b`");
+
+assert (cr.length === 3
+        && lf.length === 3
+        && crlf.length === 3
+        && cr[1] === lf[1]
+        && lf[1] === crlf[1]
+        && crlf[1] === '\n');
```
Comment (zherczeg) [inline tests/jerry/es2015/template_string.js]: I would use separate asserts.

File: tests/jerry/es2015/tagged-template-literal.js [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+  let str = (arr) => arr;
+  function getStr() {
+    return str`foo`;
+  }
+  var chainedCall = getStr();
+  var localNew = new getStr();
+  assert(chainedCall === getStr() && chainedCall === localNew);
+})();
+
+
```
Comment (zherczeg) [inline tests/jerry/es2015/tagged-template-literal.js]: One newline.

**Comment by rerobika** [comment]
@zherczeg Thanks for the review, the PR is updated.

File: jerry-core/ecma/base/ecma-gc.c [hunk 12]
```diff
@@ -362,6 +363,22 @@ ecma_gc_mark_container_object (ecma_object_t *object_p) /**< object */
...
 
+/**
+ * Mark tagged template literals of the compiled code.
+ */
+static void
+ecma_gc_mark_tagged_template_literals (const ecma_compiled_code_t *byte_code_p)
+{
+  JERRY_ASSERT (byte_code_p->status_flags & CBC_CODE_FLAG_HAS_TAGGED_LITERALS);
+
+  ecma_collection_t *coll_p = ecma_compiled_code_get_tagged_template_collection (byte_code_p);
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-gc.c]: collection_p

File: jerry-core/ecma/base/ecma-helpers.h [hunk 13]
```diff
@@ -425,6 +425,12 @@ void ecma_raise_error_from_error_reference (ecma_value_t value);
 
 void ecma_bytecode_ref (ecma_compiled_code_t *bytecode_p);
 void ecma_bytecode_deref (ecma_compiled_code_t *bytecode_p);
+#if ENABLED (JERRY_LINE_INFO) || ENABLED (JERRY_ES2015_MODULE_SYSTEM) || ENABLED (JERRY_ES2015)
+ecma_length_t ecma_compiled_code_get_formal_params (const ecma_compiled_code_t *bytecode_p);
+#endif /* ENABLED (JERRY_LINE_INFO) || ENABLED (JERRY_ES2015_MODULE_SYSTEM) || ENABLED (JERRY_ES2015) */
+#if ENABLED (JERRY_ES2015)
+ecma_collection_t *ecma_compiled_code_get_tagged_template_collection (const ecma_compiled_code_t *bytecode_header_p);
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers.h]: These are in reversed order in the C file.

File: jerry-core/ecma/operations/ecma-function-object.c [hunk 14]
```diff
@@ -45,30 +45,21 @@ ecma_op_resource_name (const ecma_compiled_code_t *bytecode_header_p)
...
-      formal_params_number = args_p->argument_end;
-    }
-  }
+  ecma_length_t formal_params_number = ecma_compiled_code_get_formal_params (bytecode_header_p);
 
   uint8_t *byte_p = (uint8_t *) bytecode_header_p;
   byte_p += ((size_t) bytecode_header_p->size) << JMEM_ALIGNMENT_LOG;
 
   ecma_value_t *resource_name_p = (ecma_value_t *) byte_p;
   resource_name_p -= formal_params_number;
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-function-object.c]: get `ecma_compiled_code_get_formal_params (bytecode_header_p);` here. No need local variables.

File: jerry-core/parser/js/js-parser-expr.c [hunk 15]
```diff
@@ -1110,7 +1111,6 @@ parser_parse_function_expression (parser_context_t *context_p, /**< context */
 } /* parser_parse_function_expression */
 
 #if ENABLED (JERRY_ES2015)
-
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-expr.c]: No need to remove this.

File: jerry-core/parser/js/js-parser-expr.c [hunk 16]
```diff
@@ -1189,30 +1190,93 @@ parser_parse_template_literal (parser_context_t *context_p) /**< context */
...
+    if (collection_p->item_count > CBC_MAXIMUM_BYTE_VALUE)
+    {
+      parser_raise_error (context_p, PARSER_ERR_ARGUMENT_LIMIT_REACHED);
+    }
+  }
+
+  const uint32_t tagged_id = collection_p->item_count;
+  uint32_t prop_idx = 0;
+  ecma_object_t *raw_strings_p;
+  ecma_object_t *template_obj_p = parser_new_tagged_template_literal (&raw_strings_p);
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-expr.c]: Do we need this helper? Looks simple, and used only once.

File: jerry-core/parser/js/js-parser-tagged-template-literal.c [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+ */
+void
+parser_tagged_template_literal_finalize (ecma_object_t *template_obj_p, /**< template object */
+                                         ecma_object_t *raw_strings_p) /**< raw strings object */
+{
+  parser_tagged_template_literal_freeze_array (template_obj_p);
+  parser_tagged_template_literal_freeze_array (raw_strings_p);
+
+  ecma_deref_object (template_obj_p);
+} /* parser_tagged_template_literal_finalize */
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-tagged-template-literal.c]: This is also simple and used only once.

File: jerry-core/parser/js/js-parser-expr.c (hunk 16 above)
Comment (rerobika) [reply jerry-core/parser/js/js-parser-expr.c]: I'd prefer these helpers functions. I think it helps to separate the runtime functionality from the parse time also the whole property management is performed here.

File: jerry-core/parser/js/js-parser-mem.c [hunk 18]
```diff
@@ -82,12 +82,27 @@ parser_malloc_local (parser_context_t *context_p, /**< context */
 /**
  * Free memory allocated by parser_malloc_local.
  */
-void parser_free_local (void *ptr, /**< pointer to free */
+void
+parser_free_local (void *ptr, /**< pointer to free */
                         size_t size) /**< size of the memory */
```
Comment (dbatyai) [inline jerry-core/parser/js/js-parser-mem.c]: Argument is misaligned.

Comment (rerobika) [reply jerry-core/parser/js/js-parser-mem.c]: Thanks, fixed.
